src/main/java/org/apache/sling/resourceresolver/impl/FactoryRegistrationHandler.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.sling.resourceresolver.impl;
+
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.apache.sling.api.resource.runtime.RuntimeService;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.Constants;
+import org.osgi.framework.ServiceFactory;
+import org.osgi.framework.ServiceRegistration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Dictionary;
+import java.util.Hashtable;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public class FactoryRegistrationHandler implements AutoCloseable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FactoryRegistrationHandler.class);
+
+    private final ResourceResolverFactoryActivator activator;
+
+    private final FactoryPreconditions factoryPreconditions;
+
+    private final ExecutorService factoryRegistrationWorker;
+
+    @SuppressWarnings("java:S3077") // The field is only ever set to null or to a new FactoryRegistration instance, which is safe.
+    private volatile FactoryRegistration factoryRegistration;
+
+    public FactoryRegistrationHandler(
+            ResourceResolverFactoryActivator activator, FactoryPreconditions factoryPreconditions) {
+        this.factoryPreconditions = factoryPreconditions;
+        this.activator = activator;
+        this.factoryRegistrationWorker = Executors.newSingleThreadExecutor(
+                r -> new Thread(r, "ResourceResolverFactory registration/deregistration"));
+    }
+
+    @Override
+    public void close() {
+        unregisterFactory();
+        factoryRegistrationWorker.shutdown();
+        try {
+            if (!factoryRegistrationWorker.awaitTermination(5, TimeUnit.SECONDS)) {
+                final List<Runnable> runnables = factoryRegistrationWorker.shutdownNow();
+                if (runnables.size() >= 2) {
```

Comment:
This deserves a comment/method, I can't tell why we check for more than two runnables or why we ignore the first two.

Reply:
I think you found an artifact from a previous iteration of the patch. The intention was to make sure that the RRF is unregistered before we exit the `close()` method. I'll fix the code accordingly.

Reply:
Fixed with https://github.com/apache/sling-org-apache-sling-resourceresolver/pull/100/commits/64a00364721b44d9319219ce9ac0e102fa8032fa

Reply:
Thanks!